### PR TITLE
[WIP] HBase connector

### PR DESCRIPTION
**Pull request by damiencarol**
This is a first version. (naive mapping Cells<->JSON message)
Most of stuff works (metadata, columns definitions, scan, ...)

More commit will comes.

### CODE REVIEW

File: presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseSimpleConsumerManager.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+    private class SimpleConsumerCacheLoader
+            extends CacheLoader<HostAddress, Connection>
+    {
+        @Override
+        public Connection load(HostAddress host)
+                throws Exception
+        {
+        	Connection connection = ConnectionFactory.createConnection(config);
+            log.info("Creating new Connection for %s", host);
```
Comment (kokosing) [inline presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseSimpleConsumerManager.java]: I would put logging into `ConnectionFactory.createConnection` (or drop).

File: presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseSimpleConsumerManager.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+            return connection;
+            /*return new Connection(host.getHostText(),
+                    host.getPort(),
+                    connectTimeoutMillis,
+                    bufferSizeBytes,
+                    format("presto-hbase-%s-%s", connectorId, nodeManager.getCurrentNode().getNodeIdentifier()));*/
```
Comment (kokosing) [inline presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseSimpleConsumerManager.java]: remove commented code

File: presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseSplitManager.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    	} catch (IOException ex) {
+    		log.error(ex.getMessage(), ex);
+    	}
+        
+        return new FixedSplitSource(splits.build());
+    }
+
+    private static <T> T selectRandom(Iterable<T> iterable)
+    {
+        List<T> list = ImmutableList.copyOf(iterable);
```
Comment (kokosing) [inline presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseSplitManager.java]: maybe store nodes as list, so you would not need to copy them each time you find splits

File: presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseSplitManager.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	
+	        Connection connection = consumerManager.getConsumer(selectRandom(nodes));
+	        
+	        RegionLocator regionLocator=  connection.getRegionLocator(tableName);
+	        List<HRegionLocation> regionLocationList = regionLocator.getAllRegionLocations();
+	
+	
+	        
+	        for (HRegionLocation regionLocation : regionLocationList) {
+	        	HRegionInfo regionInfo = regionLocation.getRegionInfo();
```
Comment (kokosing) [inline presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseSplitManager.java]: format (indentation) this file

File: presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseTableColumnGroup.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+    @Override
+    public String toString()
+    {
+        return toStringHelper(this)
+                .add("dataFormat", dataFormat)
+                .add("fields", fields)
+                .toString();
+    }
+}
```
Comment (kokosing) [inline presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseTableColumnGroup.java]: no new line

File: presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseTableDescriptionSupplier.java [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+            log.debug("Loaded Table definitions: %s", tableDefinitions.keySet());
+
+            builder = ImmutableMap.builder();
+            for (String definedTable : tableNames) {
+                SchemaTableName tableName;
+                try {
+                    tableName = SchemaTableName.valueOf(definedTable);
+                }
+                catch (IllegalArgumentException iae) {
```
Comment (kokosing) [inline presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseTableDescriptionSupplier.java]: it looks like a hack. Can you replace `try-catch` with `if`?

File: presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseRecordCursor.java [hunk 7]
```diff
@@ -0,0 +1,256 @@
+package com.facebook.presto.hbase;
```
Comment (kokosing) [inline presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseRecordCursor.java]: no copyrights

File: presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseRecordCursor.java [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+	private void checkFieldType(int field, Class<?> expected) {
+		Class<?> actual = getType(field).getJavaType();
+		checkArgument(actual == expected, "Expected field %s to be type %s but is %s", field, expected, actual);
+	}
+
+	@Override
+	public void close() {
+	}
+
+}
```
Comment (kokosing) [inline presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseRecordCursor.java]: no new line

File: presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseRecordCursor.java [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+import com.facebook.presto.decoder.RowDecoder;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.RecordCursor;
+import com.facebook.presto.spi.type.Type;
+
+import io.airlift.log.Logger;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+
+public class HBaseRecordCursor implements RecordCursor {
```
Comment (kokosing) [inline presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseRecordCursor.java]: how it is extracted where no there is no deletion lines.

File: presto-hbase/pom.xml [hunk 10]
```diff
@@ -0,0 +1,146 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (damiencarol) [inline presto-hbase/pom.xml]: Will address it in next push.

File: presto-hbase/pom.xml [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-lang3</artifactId>
+            <version>3.1</version>
+        </dependency>
+
+		<dependency>
+			<groupId>org.apache.hbase</groupId>
+			<artifactId>hbase-client</artifactId>
+			<version>${hbase.version}</version>
+		</dependency>
```
Comment (damiencarol) [inline presto-hbase/pom.xml]: Will address it in next push.

File: presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseColumnHandle.java [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+ */
+public final class HBaseColumnHandle
+        implements DecoderColumnHandle, Comparable<HBaseColumnHandle>
+{
+    private final String connectorId;
+    private final int ordinalPosition;
+
+    /**
+     * Column Name
+     */
```
Comment (damiencarol) [inline presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseColumnHandle.java]: Copy/paste from kafka. Will address it in next push.

File: presto-hbase/src/main/java/com/facebook/presto/hbase/HBasePlugin.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+ * Presto plugin to use Apache HBase as a data source.
+ */
+public class HBasePlugin
+        implements Plugin
+{
+    private Optional<Supplier<Map<SchemaTableName, HBaseTableDescription>>> tableDescriptionSupplier = Optional.empty();
+    private Map<String, String> optionalConfig = ImmutableMap.of();
+
+    @Override
+    public synchronized void setOptionalConfig(Map<String, String> optionalConfig)
```
Comment (losipiuk) [inline presto-hbase/src/main/java/com/facebook/presto/hbase/HBasePlugin.java]: make it not synchronized

File: presto-hbase/src/main/java/com/facebook/presto/hbase/HBasePlugin.java [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    @VisibleForTesting
+    public synchronized void setTableDescriptionSupplier(Supplier<Map<SchemaTableName, HBaseTableDescription>> tableDescriptionSupplier)
+    {
+        this.tableDescriptionSupplier = Optional.of(requireNonNull(tableDescriptionSupplier, "tableDescriptionSupplier is null"));
+    }
+
+    @Override
+    public synchronized Iterable<ConnectorFactory> getConnectorFactories(ConnectorFactoryContext context)
```
Comment (losipiuk) [inline presto-hbase/src/main/java/com/facebook/presto/hbase/HBasePlugin.java]: make it not synchronized

File: presto-hbase/src/main/java/com/facebook/presto/hbase/HBasePlugin.java [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+    private Map<String, String> optionalConfig = ImmutableMap.of();
+
+    @Override
+    public synchronized void setOptionalConfig(Map<String, String> optionalConfig)
+    {
+        this.optionalConfig = ImmutableMap.copyOf(requireNonNull(optionalConfig, "optionalConfig is null"));
+    }
+
+    @VisibleForTesting
+    public synchronized void setTableDescriptionSupplier(Supplier<Map<SchemaTableName, HBaseTableDescription>> tableDescriptionSupplier)
```
Comment (losipiuk) [inline presto-hbase/src/main/java/com/facebook/presto/hbase/HBasePlugin.java]: Also I do not belive synchronized would be needed here.

File: presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseRecordCursor.java [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+
+	private static final Logger log = Logger.get(HBaseRecordCursor.class);
+
+    //private static final int HBASE_READ_BUFFER_SIZE = 100_000;
```
Comment (losipiuk) [inline presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseRecordCursor.java]: remove

File: presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseRecordCursor.java [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+	@Override
+	public Type getType(int field) {
+		checkArgument(field < columnHandles.size(), "Invalid field index");
+		return columnHandles.get(field).getType();
+	}
+
+	@Override
+	public boolean advanceNextPosition() {
+		try
+		{
```
Comment (losipiuk) [inline presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseRecordCursor.java]: this `{` should be in previous line

File: presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseRecordCursor.java [hunk 18]
```diff
@@ -0,0 +1,256 @@
...
+				return endOfData(); // Past our split end. Bail.
+			}
+		} catch (IOException ex) {
+			throw new PrestoException(HBASE_CURSOR_ERROR, "could not get next from HBase scanner, error is '" + ex.getMessage() + "'");
+		}
+	}
+
+	private boolean endOfData() {
+		if (!reported.getAndSet(true)) {
+			log.debug("Found a total of %d messages with %d bytes", totalMessages, totalBytes);
```
Comment (losipiuk) [inline presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseRecordCursor.java]: copy paste from kafka

File: presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseInternalFieldDescription.java [hunk 19]
```diff
@@ -0,0 +1,281 @@
...
+    @Override
+    public String toString()
+    {
+        return toStringHelper(this)
+                .add("name", name)
+                .add("type", type)
+                .toString();
+    }
+
+    public class BooleanHBaseFieldValueProvider
```
Comment (losipiuk) [inline presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseInternalFieldDescription.java]: I would make these inner methods `static` and pass `name` explicitly to instances via constructor.

File: presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseRecordCursor.java [hunk 20]
```diff
@@ -0,0 +1,256 @@
...
+			messageData = new byte[message.remaining()];
+			message.get(messageData);
+		}
+
+		Set<FieldValueProvider> fieldValueProviders = new HashSet<>();
+
+		fieldValueProviders.addAll(globalInternalFieldValueProviders);
+		fieldValueProviders.add(HBaseInternalFieldDescription.SEGMENT_COUNT_FIELD.forLongValue(totalMessages));
+		fieldValueProviders.add(HBaseInternalFieldDescription.PARTITION_OFFSET_FIELD
+				.forLongValue(0 /* messageAndOffset.offset() */));
```
Comment (losipiuk) [inline presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseRecordCursor.java]: be consistent with either wrapping or not wrapping lines here.

File: presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseRecordCursor.java [hunk 21]
```diff
@@ -0,0 +1,256 @@
...
+		fieldValueProviders.add(HBaseInternalFieldDescription.MESSAGE_FIELD.forByteValue(messageData));
+		fieldValueProviders.add(HBaseInternalFieldDescription.MESSAGE_LENGTH_FIELD.forLongValue(messageData.length));
+		fieldValueProviders.add(HBaseInternalFieldDescription.KEY_FIELD.forByteValue(messageAndOffset.getRow()));
+		fieldValueProviders
+				.add(HBaseInternalFieldDescription.KEY_LENGTH_FIELD.forLongValue(messageAndOffset.getRow().length));
+
+		fieldValueProviders.add(HBaseInternalFieldDescription.KEY_CORRUPT_FIELD.forBooleanValue(true));
```
Comment (losipiuk) [inline presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseRecordCursor.java]: always true?

File: presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseRecordCursor.java [hunk 22]
```diff
@@ -0,0 +1,256 @@
...
+        fieldValueProviders.add(HBaseInternalFieldDescription.MESSAGE_CORRUPT_FIELD.forBooleanValue(true));
```
Comment (losipiuk) [inline presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseRecordCursor.java]: ditto?

File: presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseSplit.java [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+ * Represents a HBase specific {@link ConnectorSplit}. Each split is mapped to a segment file on disk (based off the segment offset start() and end() values) so that
+ * a partition can be processed by reading segment files from partition leader. Otherwise, a HBase topic could only be processed along partition boundaries.
+ * <p/>
+ * When planning to process a HBase topic with Presto, using smaller than the recommended segment size (default is 1G) allows Presto to optimize early and process a topic
+ * with more workers in parallel.
+ */
+public class HBaseSplit
+        implements ConnectorSplit
+{
+    private final String connectorId;
```
Comment (losipiuk) [inline presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseSplit.java]: Use `Optional` for optional fields.

File: presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseRecordSet.java [hunk 24]
```diff
@@ -0,0 +1,355 @@
...
+        return columnTypes;
+    }
+
+    @Override
+    public RecordCursor cursor()
+    {
+        return new HBaseRecordCursor();
+    }
+
+    public class HBaseRecordCursor
```
Comment (losipiuk) [inline presto-hbase/src/main/java/com/facebook/presto/hbase/HBaseRecordSet.java]: This is also defined as toplevel class

**Comment by linqerli** [comment]
what is the process of hbase connector?

**Comment by damiencarol** [comment]
@linqerli I don't understand your question. I'm working on it that's all.
A v2 will follow soon.

**Comment by rgs1** [comment]
@damiencarol are you still planning on publishing v2 of this?

**Comment by spragues-trulia** [comment]
@damiencarol - lots of great work you started... what happened?

**Comment by utf7** [comment]
@damiencarol stilling planning on supporting hbase?

**Comment by Bennyelg** [comment]
Any News??

**Comment by ideal-hp** [comment]
2018.5

**Comment by ideal-hp** [comment]
@damiencarol  I am also working hard See:
https://github.com/ideal-hp/presto-runners/tree/master/presto-hbase

**Comment by tooptoop4** [comment]
@damiencarol pls dont leave us

**Comment by userkdg** [comment]
@damiencarol I coming from June 2023.

**Pull request closed without merging**